src/vmspawn/vmspawn-scope.c

```diff
@@ -0,0 +1,426 @@
+        if (r < 0)
+                return bus_log_create_error(r);
+
+        r = sd_bus_message_open_container(m, 'v', "a(sasb)");
+        if (r < 0)
+                return bus_log_create_error(r);
+
+        r = sd_bus_message_open_container(m, 'a', "(sasb)");
+        if (r < 0)
+                return bus_log_create_error(r);
+
+        FOREACH_ARRAY(cmd, commands, n_commands) {
+                char **cmdline = *cmd;
+
+                r = sd_bus_message_open_container(m, 'r', "sasb");
+                if (r < 0)
+                        return bus_log_create_error(r);
+
+                r = sd_bus_message_append(m, "s", cmdline[0]);
+                if (r < 0)
+                        return bus_log_create_error(r);
+
+                r = sd_bus_message_append_strv(m, cmdline);
+                if (r < 0)
+                        return bus_log_create_error(r);
+
+                r = sd_bus_message_append(m, "b", 0);
+                if (r < 0)
+                        return bus_log_create_error(r);
+
+                r = sd_bus_message_close_container(m);
+                if (r < 0)
+                        return bus_log_create_error(r);
+        }
+
+        r = sd_bus_message_close_container(m);
+        if (r < 0)
+                return bus_log_create_error(r);
+
+        r = sd_bus_message_close_container(m);
+        if (r < 0)
+                return bus_log_create_error(r);
+
+        r = sd_bus_message_close_container(m);
+        if (r < 0)
+                return bus_log_create_error(r);
+
+        return 0;
+}
+
+/* NOTE: this function should really take cmdline as `const char * const *` but cannot until sd_bus_message_append_str does as well */
+int run_command_bound_to_scope(
+                sd_bus *bus,
+                const char *scope,
+                const char *service_name,
+                char **cmdline,
+                char **cleanup_cmdline,
+                char **extra_properties) {
+        _cleanup_(bus_wait_for_jobs_freep) BusWaitForJobs *w = NULL;
+        _cleanup_(sd_bus_error_free) sd_bus_error error = SD_BUS_ERROR_NULL;
+        _cleanup_(sd_bus_message_unrefp) sd_bus_message *m = NULL, *reply = NULL;
+        _cleanup_free_ char *description = NULL, *unit_name = NULL;
+        const char *object;
+        int r;
+
+        /*
+         * Runs a command in transient service unit bound to the given scope.
+         *
+         * `cmdline` is forwarded to the unit's ExecStart property.
+         * `cleanup_cmdline` is forwarded to the unit's ExecStopPost property.
+         *
+         * `extra_properties` is used to set any extra properties this unit should have.
+         * The format is the same as in unit files, e.g. Requires=other-unit.socket
+         */
+
+        assert(bus);
+        assert(service_name);
+        assert(scope);
+        assert(cmdline);
+
+        r = unit_name_mangle_with_suffix(service_name, "as service", 0, ".service", &unit_name);
```

Comment:
the "mangle" api is primarily there to turn slightly icky user input into valid unit names. but when we create unit names internally from our own components, then these should be clean, and not require mangling. hence, please use strjoin() and friends.